MFVNeutralino/plugins/Vertexer.cc

```diff
@@ -1,891 +1,2046 @@
+        if (verbose) {
+          printf("from tracks");
+          for (auto itk : itks)
+            printf(" %lu", itk);
+          printf(": vertex #%3lu: chi2/dof: %7.3f dof: %7.3f pos: <%7.3f, %7.3f, %7.3f>  rho: %7.3f  phi: %7.3f  r: %7.3f\n", vertices->size() - 1, vchi2, vndof, vx, vy, vz, rho, phi, r);
+        }
+        if (histos) {
+          for (auto it = v.tracks_begin(), ite = v.tracks_end(); it != ite; ++it)
+            h_seed_vertex_track_weights->Fill(v.trackWeight(*it));
+          h_seed_vertex_chi2->Fill(vchi2);
+          h_seed_vertex_ndof->Fill(vndof);
+          h_seed_vertex_x->Fill(vx);
+          h_seed_vertex_y->Fill(vy);
+          h_seed_vertex_rho->Fill(rho);
+          h_seed_vertex_phi->Fill(phi);
+          h_seed_vertex_z->Fill(vz);
+          h_seed_vertex_r->Fill(r);
+        }
+      }
+    }
+  };
+
+  // ha
+  for (size_t itk = 0; itk < ntk; ++itk) {
+    itks[0] = itk;
+    for (size_t jtk = itk + 1; jtk < ntk; ++jtk) {
+      itks[1] = jtk;
+      if (n_tracks_per_seed_vertex == 2) { try_seed_vertex(); continue; }
+      for (size_t ktk = jtk + 1; ktk < ntk; ++ktk) {
+        itks[2] = ktk;
+        if (n_tracks_per_seed_vertex == 3) { try_seed_vertex(); continue; }
+        for (size_t ltk = ktk + 1; ltk < ntk; ++ltk) {
+          itks[3] = ltk;
+          if (n_tracks_per_seed_vertex == 4) { try_seed_vertex(); continue; }
+          for (size_t mtk = ltk + 1; mtk < ntk; ++mtk) {
+            itks[4] = mtk;
+            try_seed_vertex();
+          }
+        }
+      }
+    }
+  }
+
+  if (histos) {
+    for (std::vector<reco::Vertex>::const_iterator v0 = vertices->begin(); v0 != vertices->end(); ++v0) {
+      const double v0x = v0->position().x() - bsx;
+      const double v0y = v0->position().y() - bsy;
+      const double phi0 = atan2(v0y, v0x);
+      for (std::vector<reco::Vertex>::const_iterator v1 = v0 + 1; v1 != vertices->end(); ++v1) {
+        const double v1x = v1->position().x() - bsx;
+        const double v1y = v1->position().y() - bsy;
+        const double phi1 = atan2(v1y, v1x);
+        h_seed_vertex_paird2d->Fill(mag(v0x - v1x, v0y - v1y));
+        h_seed_vertex_pairdphi->Fill(fabs(reco::deltaPhi(phi0, phi1)));
+      }
+    }
+  }
+
+  if (verbose)
+    printf("n_seed_vertices: %lu\n", vertices->size());
+  if (histos)
+    h_n_seed_vertices->Fill(vertices->size());
+
+  //////////////////////////////////////////////////////////////////////
+  // Take care of track sharing. If a track is in two vertices, and
+  // the vertices are "close", refit the tracks from the two together
+  // as one vertex. If the vertices are not close, keep the track in
+  // the vertex to which it is "closer".
+  //////////////////////////////////////////////////////////////////////
+
+  if (verbose)
+    printf("fun time!\n");
+
+  track_set discarded_tracks;
+  int n_resets = 0;
+  int n_onetracks = 0;
+  std::vector<reco::Vertex>::iterator v[2];
+  std::vector<reco::Vertex>::iterator nv[2];

```

Comment:
I'm a bit confused about nv... I think it's only used for printouts, is that right? Maybe we just need a comment on it. But I'm unsure why we even need nv declared all the way up here in order to do those printouts (e.g. [here](https://github.com/DisplacedVertices/cmssw-usercode/pull/5/files?w=1#diff-eaef89c993d668ab99b433c8e0844dbd1b7a8260a673b2d83468451a634fa1caR1257-R1265).

Reply:
This nv[2] vector is different from nv dummy for each vertex in new_vertices though but you're right that They serve for the same purpose, printouts. Here are the only usage of this nv[2] after tight merging routine is done:
https://github.com/DisplacedVertices/cmssw-usercode/blob/0dcb7022d40fd368fe39dae00b241c0e1e8fedd9/MFVNeutralino/plugins/Vertexer.cc#L1240-L1245

Reply:
Thanks! It looked like it was the only place it was used to me too. The variable name was a bit confusing since `nv` was used in other places as well, e.g. [here](https://github.com/DisplacedVertices/cmssw-usercode/blob/0dcb7022d40fd368fe39dae00b241c0e1e8fedd9/MFVNeutralino/plugins/Vertexer.cc#L842).

Anyway, I reorganized things in [this commit](https://github.com/DisplacedVertices/cmssw-usercode/commit/63791cd87f636149ac2c67e279f28ae650a0bc74), since _this_ `nv` was only used for the printouts. But if anything looks wrong, just shout!